lib/puppet/application/agent.rb

```diff
@@ -358,11 +358,22 @@ def run_command
   end
 
   def fingerprint
-    sm = Puppet::SSL::StateMachine.new(onetime: true)
-    ssl_context = sm.ensure_client_certificate
-    puts Puppet::SSL::Digest.new(options[:digest].to_s, ssl_context.client_cert.to_der).to_s
-  rescue
-    $stderr.puts _("Fingerprint asked but no certificate nor certificate request have yet been issued")
+    Puppet::Util::Log.newdestination(:console)
+    cert_provider = Puppet::X509::CertProvider.new
+    client_cert = cert_provider.load_client_cert(Puppet[:certname])
+    if client_cert
+      puts Puppet::SSL::Digest.new(options[:digest].to_s, client_cert.to_der).to_s
+    else
+      csr = cert_provider.load_request(Puppet[:certname])
+      if csr
+        puts Puppet::SSL::Digest.new(options[:digest].to_s, csr.to_der).to_s
```

Comment:
I know that this is probable expected to be able to go into some kind of scripting, so this may not be possible, but it seems like we should inform the user if the fingerprint is from the cert or the CSR.  Has this been discussed before?  Maybe we should implement a `puppet ssl status` subcommand that prints several details but does not try to change state.

Reply:
Yeah it's a little surprising to me that the (old) behavior doesn't say which one the digest is from. We could append something to the output to distinguish what it is. @Magisus Does `puppetserver-ca` print request and certs fingerprints in the same action?

Reply:
Thinking about this some more, I'm going to merge this as it restores the old/existing behavior. If we want to change the behavior, and clarify whether it's the cert or request, then I think that's a new ticket